aws_lambda_powertools/utilities/idempotency/persistence/base.py

```diff
@@ -0,0 +1,432 @@
+"""
+Persistence layers supporting idempotency
+"""
+
+import datetime
+import hashlib
+import json
+import logging
+from abc import ABC, abstractmethod
+from typing import Any, Dict
+
+import jmespath
+
+from aws_lambda_powertools.utilities.idempotency.cache_dict import LRUDict
+from aws_lambda_powertools.utilities.idempotency.exceptions import (
+    IdempotencyValidationerror,
+    InvalidStatusError,
+    ItemAlreadyExistsError,
+)
+
+logger = logging.getLogger(__name__)
+
+STATUS_CONSTANTS = {
+    "NOTEXISTING": "DOESNOTEXIST",
+    "INPROGRESS": "INPROGRESS",
+    "COMPLETED": "COMPLETED",
+    "EXPIRED": "EXPIRED",
+    "ERROR": "ERROR",
+}
+
+
+class DataRecord:
+    """
+    Data Class for idempotency records.
+    """
+
+    def __init__(
+        self,
+        idempotency_key,
+        status: str = "",
+        expiry_timestamp: int = None,
+        response_data: str = "",
```

Comment:
@cakepietoast  i think we should give the option pretty early on the also save the request (or parts of it)

For example you might need some of the original request to cancel an idempotent request (https://developer.squareup.com/reference/square/payments-api/cancel-payment-by-idempotency-key)